File: core/audiosheet/symbolic/quantize.py
"""S5.2 rhythmic quantisation — ARCHITECTURE.md Section 2.1.3.

A dynamic program over grid positions, not nearest-neighbour snapping: naive
snapping produces zero-length notes, chord de-synchronisation and impossible
tuplet soups.

This module and ``packages/engine/src/quantize.ts`` implement the same DP for
different runtimes (S5 next to the models, S7 next to the slider). They MUST
agree on every vector in ``core/tests/golden/quantize/`` — that shared suite is
what stops the two implementations drifting.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Final

from audiosheet.schema import DifficultyProfile, Note, TimingGrid

#: Search half-width around a raw onset, in beats. Bounds the DP to O(N*M*W).
SEARCH_WINDOW_BEATS: Final[float] = 1.0


@dataclass(frozen=True)
class QuantizeResult:
    """The outcome of quantising one voice.

    Attributes:
        notes: The quantised notes.
        mean_shift_ticks: Mean absolute snap applied, for diagnostics.
        dropped_ids: Notes dropped for falling below the minimum duration.
    """

    notes: list[Note]
    mean_shift_ticks: float
    dropped_ids: list[str]


def quantize(
    notes: list[Note],
    grid: TimingGrid,
    profile: DifficultyProfile,
) -> QuantizeResult:
    """Snap onsets then durations to the profile's grid via the Section 2.1.3 DP.

    Ties break toward the earlier grid position, which keeps the result
    deterministic (INV-2).

    Args:
        notes: Notes of a single voice, in onset order.
        grid: The timing grid.
        profile: Difficulty profile supplying the grid and the cost weights.

    Returns:
        The quantised notes plus diagnostics.

    Raises:
        NotImplementedError: Phase 6.
    """
    raise NotImplementedError("S5.2 quantisation lands in Phase 6")


def syncopation_penalty(tick: int, grid: TimingGrid) -> float:
    """Return the off-beat penalty for a grid position, scaled by metric depth.

    Args:
        tick: Candidate grid position.
        grid: The timing grid.

    Returns:
        A non-negative penalty; 0 on a downbeat.

    Raises:
        NotImplementedError: Phase 6.
    """
    raise NotImplementedError("syncopation scoring lands in Phase 6")

File: packages/engine/src/quantize.ts
/**
 * Rhythmic quantisation — ARCHITECTURE.md Section 2.1.3.
 *
 * A dynamic program over grid positions, solved with Viterbi and bounded to
 * `SEARCH_WINDOW_BEATS` around each raw onset. Ties break toward the earlier grid
 * position, which keeps the result deterministic (INV-2).
 *
 * This module and `core/audiosheet/symbolic/quantize.py` implement the same DP
 * for different runtimes. They MUST agree on every vector in
 * `core/tests/golden/quantize/`; that shared suite is what stops them drifting.
 */

import type { DifficultyProfile, Note, TimingGrid } from "@audiosheet/schema";

/** Search half-width around a raw onset, in beats. */
export const SEARCH_WINDOW_BEATS = 1.0;

/** The outcome of quantising one voice. */
export interface QuantizeResult {
  readonly notes: Note[];
  readonly meanShiftTicks: number;
  readonly droppedIds: readonly string[];
}

/**
 * Snap onsets, then durations, to the profile's grid.
 *
 * @param notes - Notes of a single voice, in onset order.
 * @param grid - The timing grid.
 * @param profile - Supplies the grid divisions and the cost weights.
 * @throws Always in Phase 0; lands in Phase 2.
 */
export function quantize(
  notes: readonly Note[],
  grid: TimingGrid,
  profile: DifficultyProfile,
): QuantizeResult {
  throw new Error("unimplemented: quantisation lands in Phase 2");
}

/**
 * Return the legal onset lattice for a profile, in ticks.
 *
 * @param grid - The timing grid.
 * @param profile - Supplies `grid_divisions` and `allow_tuplets`.
 * @throws Always in Phase 0; lands in Phase 2.
 */
export function legalPositions(grid: TimingGrid, profile: DifficultyProfile): number[] {
  throw new Error("unimplemented: grid construction lands in Phase 2");
}
